### ISSUE TITLE
[css-view-transitions-2] Decide on naming for navigation-triggered view-transition rule

### ISSUE THREAD

**Issue by noamr** [w3c/csswg-drafts#9383]
For context, see [TPAC discussion](https://github.com/w3c/csswg-drafts/issues/8048#issuecomment-1719751618) as well of the rest of the conversation at #8048.

The resolution was that we would have an at-rule that defines the relationship between a navigation and a view-transition - for start, it would allow you to specify that any same-origin navigation results in triggering a view transition.
The final naming was left open.

Any naming proposal would have to:
- convey that this is about navigation-triggered view-transitions
- start as "same-origin", in a way that's extendible later to a URL, a pair of URLs, or some other form of conditioning

Current suggestions (but open for more!):

1. `@view-transition same-origin { trigger: navigation }`
2. `@view-transition-behavior same-origin { trigger: navigation }`
3. `@navigation-behavior same-origin { view-transition: trigger }`
4. `@nav-transition same-origin { behavior: trigger }`
5. `@view-transition navigation(same-origin) { behavior: trigger }`
6. `@auto-view-transition same-origin { trigger: navigation }`

My personal favorite which I've put in the [draft spec](https://github.com/w3c/csswg-drafts/pull/9382) is (3), as it allows extending in the future to other "navigation behaviors" if we so choose, while not being too generic, and tells the story of `if (navigation) { startViewTransition()}`.

**Comment by bokand** [comment]
I'll add one to the list for consideration:

* `@navigation-view-transition same-origin { behavior: trigger; }`

This remains specific to view-transitions but makes it clear it applies only to navigation-initiated and makes more sense for `same-origin` to be a qualifier.

**Comment by noamr** [comment]
> I'll add one to the list for consideration:
> 
> * `@navigation-view-transition same-origin { behavior: trigger; }`
> 
> This remains specific to view-transitions but makes it clear it applies only to navigation-initiated and makes more sense for `same-origin` to be a qualifier.

Yea, this is kind of like (4), I like it. will add to the list. thanks!

**Comment by noamr** [comment]
@fantasai @jakearchibald @tabatkins @vmpstr @khushalsagar  if you'd like to weigh in or add more options, and we can try to resolve this in one of the next WG calls.

**Comment by vmpstr** [comment]
It's between 1. and 3. for me, with a slight preference to 3.

The way I see this problem is we're either declaring something about the view transition, or we're declaring something about the navigation. 7 and 8 also blur the lines a bit and declare something about both.

My preference is to talk about navigation here, and qualify it with something like same-origin or some other mechanism in the future. After we have sufficiently defined which navigation we're targeting, inside of it we say what we'd like to happen and here's where `view-transition: trigger` comes in.

**Comment by bokand** [comment]
One thing I've pointed out elsewhere but want to record here. If we make this about the navigation then other navigation features may require adding qualifiers that view-transitions doesn't support. E.g.

```css
@navigation cross-origin {
  view-transition: trigger;
  some-other-feature-that-supports-cross-origin: enable;
}
```

In this case I'm not sure how we'd surface the fact that the VT trigger isn't effective in this case? In a view-transitions syntax we'd simply disallow non-supported qualifiers (like `cross-origin`) so they would fail at parse.

**Comment by noamr** [comment]
To make the case for (7), it's consistent with feature-specific rules like `@font-feature-values`, where the rule is specific enough and we don't try to create rules that support future unknown thing. The feature here is navigation view-transitions, so maybe that's what the name of the rule should be. Also makes feature-detection with `@supports(at-rule)` more explicit.

**Comment by khushalsagar** [comment]
My favourite option so far is `@navigation-view-transition same-origin { behavior: trigger; }`. The at-rule is about navigation triggered view transitions which is clearer with this.

**Comment by noamr** [comment]
Proposing for the WG call:
`@navigation-view-transition { behavior: trigger }` (without `same-origin`), following recent discussion at #8048.

**Comment by noamr** [comment]
@fantasai @tabatkins  since we didn't get to this in the CSSWG calls for several weeks, would you like to chime in here?

**Comment by dbaron** [comment]
One thought is that the naming likely depends in some ways on how you expect to extend this in the future.  One concern about many of the proposals here is that it feels like they have a *lot* of future extension points (argument to at rule, adding new descriptors, adding new values to existing descriptors) and it's not clear to me how the possible uses of those extension points differ.

The extension types that I see mentioned in #8048 are:
* extending to specific URL patterns, possibly cross-origin, in addition to just `same-origin`
* distinguishing based on navigation types or back/forward
* in item 8 above there also seems to be another suggestion for a future extensibility point, but I don't understand what it is.

It might be helpful for group discussion to have a summary of the possible future extension areas and where you'd expect each one to fit in.  (In particular, when I look at the above syntax options, I wonder whether there might be too many possible extension points, particularly when I see three places where there could later be multiple values but right now there's only one.)

(I'm also a little confused by the term navigation-triggered, and curious when a navigation-triggered view transition is different from a cross-document view transition.  Is one a subset of the other?)

**Comment by noamr** [comment]
Thanks @dbaron, let me try to summarize.

The extension points we see to this are what you mentioned:
- qualifying based on URL patterns
- qualifying on navigation types (back/forward/reload etc)
- extending to declare transition types. Option 8 above is simply that, it says that * means "any transition type".

To answer your other question, this is currently being triggered by cross-document navigations only, but an extension point could be enabling this for same-document navigations via the navigation API. My thought about that in particular is that it would probably not belong in CSS, but rather by some sort of JS opt-in that enables CSS navigation-triggered transitions for same-document navigaiton-API navs. It would be awkward if the author had the word `cross-document` sprinkled all over the CSS, and when we allow same-document navigations they would have to rename everything.

So in short, this is about navigation-triggered view-transitions, where *at this point* it's enabled only for cross-document.

**Comment by fantasai** [comment]
+1 to dbaron, I think we need to think about the syntax with these extensions in mind.

Wrt syntax proposals, here's a rough one:

```
@view-transition {
  navigation: same-origin;
}

@view-transition {
  navigation: same-origin;
  from: "/foo/*";
  to: "/bar/*";
  type: slide-out;
}
```

This is pulling all the selection information into declarations, which gives us a bit more syntax to play with, and will probably make more complicated expressions easier to write.

**Comment by khushalsagar** [comment]
For the upcoming naming discussion, @noamr clarified the idea on chat. Here it is with the proposed names:

```
@view-transition {
  navigation: auto;
  type: foo;
}
```

- `@view-transition` is the name of the at-rule.
- `navigation` is the descriptor with the list of navigations that should trigger a transition. The proposed values for now are `auto` and `none`. `auto` means all navigations except reload, aligning with the resolution at https://github.com/w3c/csswg-drafts/issues/8784, and `none` means no navigations. More navigation types will be introduced going forward.
- `type` is equivalent to the same parameter in startViewTransition [here](https://drafts.csswg.org/css-view-transitions-2/#dom-startviewtransitionoptions-types).
